unblob/handlers/filesystem/yaffs/utils.py

```diff
@@ -0,0 +1,574 @@
+        raise NotImplementedError
+
+    def __lt__(self, other):
+        return self.object_id < other.object_id
+
+    def __gt__(self, other):
+        return self.object_id > other.object_id
+
+    def __eq__(self, other):
+        return self.object_id == other.object_id
+
+    def __str__(self):
+        return f"{self.object_id}: {self.name}"
+
+
+@attr.define
+class YAFFS1Entry(YAFFSEntry):
+    chunks: List[YAFFS1Chunk]
+
+    def get_chunks(self) -> Iterable[YAFFS1Chunk]:
+        """Return a filtered and ordered list of chunks."""
+        # YAFFS1 chunks have a page_status field indicating
+        # whether or not the chunk is still active. We therefore
+        # filter out chunks that are inactive.
+
+        # YAFFS1 chunks have a serial number that is used to track
+        # which chunk takes precedence if two chunks have the same
+        # identifier. This is used in scenarios like power loss
+        # during a copy operation. Whenever we have two chunks with
+        # the same id, we only return the one with the highest serial.
+
+        for _, chunks in itertools.groupby(
+            sorted(
+                # filter out deleted chunks (page_status=0)
+                [chunk for chunk in self.chunks if chunk.page_status != 0x0],
+                key=lambda chunk: chunk.chunk_id,
+            )
+        ):
+            yield max(chunks, key=lambda chunk: chunk.serial)
+
+
+@attr.define
+class YAFFSFileVar:
+    file_size: int
+    stored_size: int
+    shrink_size: int
+    top_level: int
+
+
+@attr.define(kw_only=True)
+class YAFFS2Entry(YAFFSEntry):
+    chksum: int
+    yst_mode: int
+    yst_uid: int
+    yst_gid: int
+    yst_atime: int
+    yst_mtime: int
+    yst_ctime: int
+    equiv_id: int
+    yst_rdev: int
+    win_ctime: List[int]
+    win_mtime: List[int]
+    inband_shadowed_obj_id: int
+    inband_is_shrink: int
+    reserved: List[int]
+    shadows_obj: int
+    is_shrink: int
+    filehead: YAFFSFileVar
+    chunks: List[YAFFS2Chunk]
+
+    def get_chunks(self) -> Iterable[YAFFS2Chunk]:
+        """Return a filtered and ordered list of chunks."""
+        # The Yaffs2 sequence number is not the same as the Yaffs1 serial number!
+
+        # As each block is allocated, the file system's
+        # sequence number is incremented and each chunk in the block is marked with that
+        # sequence number. The sequence number thus provides a way of organising the log in
+        # chronological order.
+
+        # Since we're scanning backwards, the most recently written - and thus current - chunk
+        # matching an obj_id:chunk_id pair will be encountered first and all subsequent matching chunks must be obsolete and treated as deleted.
+
+        # note: there is no deletion marker in YAFFS2
+
+        for _, chunks in itertools.groupby(
+            sorted(self.chunks, key=lambda chunk: chunk.chunk_id)
+        ):
+            yield max(chunks, key=lambda chunk: chunk.seq_number)
```

Comment:
seq number can loop, 11 -> 00 (not sure how often happens, but could (https://yaffs.net/documents/how-yaffs-works chapter 9)

Reply:
I'm now using this construct:

```
# serial is a 2 bit, this function works since there's always at most
# two chunks with the same chunk_id at any given time
yield max(chunks, key=lambda chunk: ((chunk.serial + 1) & 3))
```